OnProfileUpdate/handler.ts

```diff
@@ -0,0 +1,200 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type, sonarjs/no-identical-functions */
+import * as t from "io-ts";
+import { pipe, flow } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import * as A from "fp-ts/ReadonlyArray";
+import * as E from "fp-ts/lib/Either";
+import * as TE from "fp-ts/lib/TaskEither";
+import * as RTE from "fp-ts/lib/ReaderTaskEither";
+import { NonNegativeInteger } from "@pagopa/ts-commons/lib/numbers";
+import { DataTableProfileEmailsRepository } from "@pagopa/io-functions-commons/dist/src/utils/unique_email_enforcement/storage";
+import { ProfileEmail } from "@pagopa/io-functions-commons/dist/src/utils/unique_email_enforcement";
+import {
+  ProfileModel,
+  Profile
+} from "@pagopa/io-functions-commons/dist/src/models/profile";
+import { generateVersionedModelId } from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model_versioned";
+import { CosmosErrors } from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model";
+import { Logger } from "@azure/functions";
+import { FiscalCode } from "../generated/backend/FiscalCode";
+
+const ProfileDocument = t.intersection([
+  ProfileEmail,
+  t.type({
+    isEmailValidated: t.literal(true),
+    version: NonNegativeInteger
+  })
+]);
+
+type ProfileDocument = t.TypeOf<typeof ProfileDocument>;
+
+interface IDependencies {
+  readonly dataTableProfileEmailsRepository: DataTableProfileEmailsRepository;
+  readonly profileModel: ProfileModel;
+  readonly logger: { readonly error: Logger["error"] };
+}
+
+// this function gets the last validated email for that fiscal code from `profile` collection
+const getPreviousValidatedEmail = (
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+) => (
+  profileModel: ProfileModel
+): TE.TaskEither<CosmosErrors, O.Option<Profile["email"]>> =>
+  pipe(
+    generateVersionedModelId<Profile, "fiscalCode">(fiscalCode, version),
+    id => profileModel.find([id, fiscalCode]),
+    TE.chain(
+      O.fold(
+        () => TE.of(O.none),
+        previousProfile =>
+          previousProfile.isEmailValidated
+            ? TE.of(O.some(previousProfile.email))
+            : pipe(
+                NonNegativeInteger.decode(version - 1),
```

Comment:
I guess we should limit this recursion somehow, otherwise a user can DoS the system making a lot of updates without a validated email causing N more queries for each update.

Reply:
what about

```
SELECT TOP 1 c.email
FROM c
WHERE c.userId = @userId
AND c.isValidatedEmail = true
ORDER BY c.versionNumber DESC
```

ensure that `isValidatedEmail` is indexed before run this and check the needed RU/s for this kind of query. Is it's too high, consider another approach.

Reply:
moreover, it looks like this method will run for each profile entry (millions currently) in the collection, so lets suppose we have a user with 100 profile versions with no email validated, this will cause 100+99+98...+1 queries (or maybe I'm missing something)